core/src/main/java/cucumber/api/TestCase.java

```diff
@@ -50,16 +55,21 @@ public TestCase(List<TestStep> testSteps, PickleEvent pickleEvent, boolean dryRu
      */
     @Deprecated
     public void run(EventBus bus) {
-        boolean skipNextStep = this.dryRun;
+        TestCase.SkipStatus skipNextStep = this.dryRun ? SkipStatus.SKIP_ALL_SKIPABLE : SkipStatus.RUN_ALL;
         Long startTime = bus.getTime();
         bus.send(new TestCaseStarted(startTime, this));
         ScenarioImpl scenarioResult = new ScenarioImpl(bus, pickleEvent);
+        boolean finishingGherkinStep = false;
         for (TestStep step : testSteps) {
+            if (skipNextStep == SkipStatus.RUN_HOOKS && finishingGherkinStep && step.startingGherkinStepType()) {
+                skipNextStep = SkipStatus.SKIP_ALL_SKIPABLE;
+            }
             Result stepResult = step.run(bus, pickleEvent.pickle.getLanguage(), scenarioResult, skipNextStep);
-            if (!stepResult.is(Result.Type.PASSED)) {
-                skipNextStep = true;
+            if (!stepResult.is(Result.Type.PASSED) && skipNextStep == SkipStatus.RUN_ALL) {
```

Comment:
I find it a bit hard to qualify why but the concept underlying this solution feels wrong. TestStep is leaking information into TestCase as to how the test should proceed. It shouldn't have to.

I think this is caused by trying to represent the test execution as a list of steps rather then the tree that it is (it may help to imagine the existence of a before step hook).

Reply:
I have one suggestion, why can't we have skipStep as binary bit flags, or bit array or a map for eg ```{"skipBeforeStep": false, "skipGherkingStep": false, "skipAfterStep": false}```

when step fails set **skipGherkingStep** to **true** but **skipAfterStep** still remains **false** and executes the afterStep. After execution of afterStep, if **skipGherkingStep** is true , set **skipAfterStep** to **true** so that none of the next afterStep executes